[FIX] web, website: adjust input colors & make disabled inputs more recognizable

Brieuc-brd opened the pull request:
### Adjust disabled input colors  (*1)
1. Maintain border consistency of frontend inputs :
Prior to this PR, `form-select` borders overlapped the background color, which is not the case with `form-control`.
This is because `form-control` uses the `background-clip` property. As we're using a semi-transparent border on frontend inputs, this creates a color issue: when a `form-select` input is disabled, the border color is darker than that of the `form-control`.
This PR adapts the `background-clip` on `form-select` input to maintain color consistency between inputs.

2. Make disabled inputs more recognizable :
Prior to this PR, disabled inputs were not sufficiently distinct from regular inputs, especially `website_sale` inputs which had a gray background.


**Steps to reproduce:**
- Make sure your instance has`website_sale_renting` installed.
- Go to the Shop page.
- Look for a product with a rental period (eg. Printer).
- Click on Add to cart, this will disable the rental period input.
- -> The gray search bar and the disabled input have almost the same style

This PR adapts the style of disabled inputs in the frontend to make them more recognizable.

---

### Adjust dropdown caret color (*2)
Prior to this PR, dropdown inputs had a color issue in frontend: the caret color of `form-select` was dark regardless of the input's background and didn't provide enough contrast when we defined a dark background on the page.

**Steps to reproduce:**
- Go to the Contact page.
- Click on Edit.
- Click on Theme tab.
- Replace the fourth color with a dark one (ex. black).
- Click on the form in the page.
- Add a field.
- Select the new field and choose the "Selection" type.
- Click on Save.
- -> The dropdown caret is not enough visible.

---

### Adjust `form-check-input` colors (*3)
This PR adjusts the caret color to make sure that this will be always visible.
Prior to this PR, checkboxes, radios and switch inputs had a color issue in frontend: as the inner element (eg. check mark) was always white if the user changed the primary color to a bright one, the inputs were no longer readable.

**Steps to reproduce:**
- Go to the Contact page.
- Click on Edit.
- Click on Theme tab.
- Replace the primary color with a bright one (ex. light gray).
- Click on the form in the page.
- Add a field.
- Select the new field and choose the "Radio Buttons" or " Checkbox" type.
- Click on Save.
- Check the checkbox or the radio button.
- -> The check mark or the dot is not visible enough.

This PR adjusts the colors to ensure that these inputs will always be visible.

---

### Adjust `form-range` track color (*4)
Prior to this PR, the `form-range` track color used the light color.
This created a color inconsistency with the rest of the UI when we changed the third color.

**Steps to reproduce:**
- Go to the Shop page.
- Click on Edit.
- Click on the page and make sure "Price Filter" is active in the Web
  Editor.
- Go to Theme tab.
- Change color-3 (light) to another one (eg. red).

This PR adjusts the color to maintain consistency with the rest of UI elements

---

###  Adjust custom dropdown caret (*5)
Prior to this PR, the custom dropdown caret in the `website_sale` sidebar didn't handle the "multiple" attribute,
unlike the default dropdown.

If we decided to add a "multiple" attribute to this element, the caret remained displayed, which created a design issue.

This PR adapts the caret of this dropdown so that it works correctly when this attribute is defined.

---


task-3702675

---
| (*1) | Before | After |
| -- | - | - |
| Portal | ![disabled_17_portal](https://github.com/odoo/odoo/assets/80679690/2c232d4d-d82d-4163-b940-b327d2234707) | ![disabled_br_portal](https://github.com/odoo/odoo/assets/80679690/1770907e-a583-4867-899f-1b6679ecb85a) |
| Website (light bg) | ![disabled_17_website_light](https://github.com/odoo/odoo/assets/80679690/f1717742-b957-4d8c-a605-8af353f33a72) | ![disabled_br_website_light](https://github.com/odoo/odoo/assets/80679690/cc622f45-4cdc-41a2-9006-aadd9e869097) |
| Website (dark bg) | ![disabled_17_website_dark](https://github.com/odoo/odoo/assets/80679690/698920f3-22f5-4a33-9dc9-a9c485b7d161) | ![disabled_br_website_dark](https://github.com/odoo/odoo/assets/80679690/4f55ceed-9760-4e20-9f2a-d108c9ae67fb) |

| (*2) | Before | After |
| -- | - | - |
| Adjust dropdown caret color | ![dropdown_17](https://github.com/odoo/odoo/assets/80679690/d46e1d3a-9d8f-403d-a3b7-22373b84e67f) | ![dropdown_br](https://github.com/odoo/odoo/assets/80679690/dcc32347-2069-42f3-b4df-abcb98a5d821) |

| (*3) | Before | After |
| -- | - | - |
| Adjust `form-check-input` colors | ![check-input_17](https://github.com/odoo/odoo/assets/80679690/510690cc-612b-428c-a16f-b236af4a8e2c) |  ![check-input_br](https://github.com/odoo/odoo/assets/80679690/88f36858-82bd-4b4c-944a-8a3a620db4c1) |

| (*4) | Before | After |
| -- | - | - |
| Adjust `form-range` track color | ![range_17](https://github.com/odoo/odoo/assets/80679690/91766309-e807-4044-beb8-9427ed982896) | ![range_br](https://github.com/odoo/odoo/assets/80679690/5c644f1d-df7f-4f6f-9efe-2e0784b93f28) |

| (*5) | Before | After |
| -- | - | - |
| Adjust custom dropdown caret | ![sale_drop_17](https://github.com/odoo/odoo/assets/80679690/97b9b4a1-d0b2-4acc-a554-05bab371c49b) | ![sale_drop_br](https://github.com/odoo/odoo/assets/80679690/28149036-0801-4711-81ee-9e7be3963c97) |





---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/150886).

Brieuc-brd commented:
Hey @stefanorigano could you take a look at this and tell me what you think ? :)

Brieuc-brd commented:
Hey @stefanorigano 👋 
I've updated the PR, could you take a look at it ?

stefanorigano reviewed (changes requested):
Thanks  @Brieuc-brd 👍 

We should address `<select>` inputs as well, I think that's within the scope of this PR.
They get emphasised on :hover, even when disabled. The root cause seems to be in `bootstrap_review.scss`, could you have a look? 

Thanks!

![image](https://github.com/odoo/odoo/assets/8558351/ea2037d7-3c3f-4cec-a23f-1dbf3f32e004)

Brieuc-brd commented:
Thanks for your feedback @stefanorigano 👍 
I've updated the branch to better handle `form-select`, what do you think ?

stefanorigano commented on addons/web/static/src/scss/bootstrap_review_frontend.scss:
```diff
@@ -167,3 +167,13 @@ $o-navbar-nav-pills-link-border-radius: $nav-pills-border-radius !default;
 }
+
+.form-select {
+    background-clip: padding-box;
```
I'd suggest to add a note explaining that's necessary in order to achieve a consistent "border rendering behaviour" across `form-select` and  `form-control`, especially when using semi-transparent borders.

stefanorigano reviewed (changes requested):
Thanks @Brieuc-brd!
Some remarks and questions for you  ;)

Brieuc-brd commented on addons/web/static/src/scss/bootstrap_review_frontend.scss:
```diff
@@ -167,3 +167,13 @@ $o-navbar-nav-pills-link-border-radius: $nav-pills-border-radius !default;
+    background-clip: padding-box;
+}
+
+.form-select:disabled, .form-control:disabled, .form-control [readonly] {
```
It was my mistake 🤦 
I updated the PR :)

This PR also handles the caret color of `form-select` now, what do you think ?

stefanorigano reviewed (changes requested):
Almost there, thanks so much @Brieuc-brd 🙌

There's just one detail missing I guess. Since checkboxes and radio [use opacity](https://github.com/twbs/bootstrap/blob/1a6fdfae6be09b09eaced8f0e442ca6f7680a61e/scss/_variables.scss#L919) to style the disabled state, they becomes almost invisible 'cause of our semi-transparent border.. could you handle them as well?

| this PR | ± desired result|
|--------|--------|
| ![image](https://github.com/odoo/odoo/assets/8558351/01fc9a7c-af38-4bdd-9afe-cdea8a06bc34) | ![image](https://github.com/odoo/odoo/assets/8558351/43ce3f37-83ef-4a32-ad9a-ddb7f7a5c505)
 |

Brieuc-brd commented:
Hey @stefanorigano 👋 
Good catch ! 
I've updated the branch, the diff :
https://github.com/odoo/odoo/compare/bba4ff710c3841723e50c2a5946f54d03a4431f9..2119b511d856093cba7b2def7f49a7d24e554f1f

stefanorigano commented on addons/web/static/src/scss/bootstrap_review_frontend.scss:
```diff
@@ -219,3 +219,25 @@ $-color-for-gray-200-bg: adjust-color-to-background($body-color, $gray-200);
+}
+
+.form-control {
+    &:disabled, [readonly] {
```
Missing parent selector  😉
```suggestion
    &:disabled, &[readonly] {
```

Brieuc-brd commented:
Hey @stefanorigano 👋 
Thanks for your feedback 🙏 

I checked again the list of inputs provided by Bootstrap and made other changes in addition to the range input.
The diffs :
- https://github.com/odoo/odoo/compare/610e036ad3f8524c81078925067d6b508a42eeff..74e02c194a20815c8d28fed698a0205a88d15ceb
- https://github.com/odoo/odoo/compare/7c8da6163beea193d4314474fdb167d1a3467f33..29c51c37c9f225b7368b0892a200da15ad38b0f4

Could you take a look at it ? :)

stefanorigano commented on addons/web/static/src/scss/bootstrap_review_frontend.scss:
```diff
@@ -219,3 +219,33 @@ $-color-for-gray-200-bg: adjust-color-to-background($body-color, $gray-200);
+
+.form-check-input:disabled:not(:checked) {
+    background-color: $input-disabled-bg;
+    opacity: 1;
```
Can't we set `$form-check-input-disabled-opacity` to `1`  instead 🤔

stefanorigano reviewed (approved):
Good @Brieuc-brd , thanks so much.
One last detail to be addressed, but it's ready for tech. review IMO.

Well done 👍

Brieuc-brd commented:
Hello @odoo/rd-website 👋 
This PR is ready for tech. review, could you take a look at it ? :)

qsm-odoo commented on addons/web/static/src/scss/bootstrap_review_frontend.scss:
```diff
@@ -219,3 +219,9 @@ $-color-for-gray-200-bg: adjust-color-to-background($body-color, $gray-200);
 }
+
+.form-select {
+    // This is necessary to achieve a consistent "border rendering behaviour"
+    // across form-select and form-control, particularly when using semi-transparent borders.
```
Nitpicking:
- Max 80 characters per line for comments (and use those 80 characters of course)
- Commit title: `frontend inputs, maintain border consistency` => `maintain border consistency of frontend inputs`

Also remind me: what's the justification for this to be part of `bootstrap_review_frontend.scss` instead of `bootstrap_review.scss`. I think I failed to see the exact point when we split... or at least I forgot now. The point of the file was to "extend bootstrap to fix it / change its behavior in Odoo". Why would not we want to fix it the same way in backend and frontend? Feel like anything that should be frontend or backend only should not be in a file that contains the word "bootstrap".

Brieuc-brd replied:
> Max 80 characters per line for comments (and use those 80 characters of course)

Done ✅ 

> Commit title: frontend inputs, maintain border consistency => maintain border consistency of frontend inputs

In the end, I've updated the branch using a single commit.

> Also remind me: what's the justification for this to be part of bootstrap_review_frontend.scss instead of bootstrap_review.scss

This fix is related to inputs with a semi-transparent border color and a background color.
If I'm not mistaken, this type of inputs is only defined in the frontend 🤔 .

I don't think we have inputs with a semi-transparent border color in the backend, or maybe I've missed something?
This is why it would make more sense to define it in `bootstrap_review_frontend.scss` rather than `bootstrap_review.scss`.

What do you think ?

cc. @stefanorigano

stefanorigano replied:
Hello 👋

@qsm-odoo , @Brieuc-brd 

> I think I failed to see the exact point when we split... or at least I forgot now. The point of the file was to "extend bootstrap to fix it / change its behavior in Odoo".

The split has been introduced in https://github.com/odoo/odoo/pull/116641 (see in the PR description, SCSS section).

In short, there are "different Odoo(s)". The webclient deviated too much from BS (just see [how](https://github.com/odoo/odoo/blob/17.0/addons/web/static/src/scss/bootstrap_review_backend.scss) we render buttons there 😅) , while the frontend must remains "BS compliant" of course.
The split is needed to finetune the behaviour for a bundle (like the frontend input border-color) while avoiding to affect of the other one (see the complex rendering for webclient buttons, definitely not needed in the frontend).

Regarding the case of disabled inputs in the webclient:
- The current approach leans towards the pragmatic "_disabled inputs are unsightly, just hide them_" (sigh 😒).
- In isolated scenarios where Bootstrap inputs are utilized (like in knowledge, mail, POS, or rooms), the colors are solid, no need for this custo

qsm-odoo commented on addons/website/static/src/scss/bootstrap_overridden.scss:
```diff
@@ -246,8 +246,15 @@ $input-placeholder-color: mix($input-bg, $input-color) !default;
+$form-range-thumb-bg: o-color('primary') !default;
+
+$form-switch-color: $input-color !default;
+$form-switch-checked-color: $form-check-input-checked-color !default;
```
Feels like some of these should in the web bootstrap_overridden, not the website one. Anything that overrides the frontend should be:
- In website if it depends on a website-only variable
- Otherwise in portal if it depends on a portal-only variable
- Otherwise in web_editor if it depends on a web_editor-only variable
- Otherwise in web if it depends on nothing / bootstrap variable (like the first one here for example: "the form select indicator is the input color" => this is not something that should be website specific.

Brieuc-brd replied:
Indeed ! I moved the variables in `web/bootstrap_overriden_frontend.scss` so that it works without installing website. 👍

qsm-odoo commented on addons/web/static/src/scss/bootstrap_review.scss:
```diff
@@ -93,7 +93,7 @@
 
 // Force field label pointer to cursor
-.form-check, .form-select {
+.form-check, .form-select:not(:disabled) {
```
Especially true in stable but in this case maybe also in master (not sure I made up my mind about that yet), I would not increase the specificity of rules that way. You can do:
```scss
.form-select:where(:not(:disabled)) {
```
It allows to target the same things but the specificity of the rule is not increasing by the "not" part. That way if someone made something like
```scss
.super-cursor {
    pointer: super;
}
```
and add it on a specific form-select, your fix here would not break his super cursor.

qsm-odoo commented on addons/web/static/src/scss/bootstrap_review_frontend.scss:
```diff
@@ -220,8 +220,31 @@ $-color-for-gray-200-bg: adjust-color-to-background($body-color, $gray-200);
     }
 }
 
+// Review disabled inputs
```
Could use some more comments than this to know what this is about.

qsm-odoo commented on addons/website/static/src/scss/bootstrap_overridden.scss:
```diff
@@ -245,13 +245,20 @@ $input-color: color-contrast($input-bg) !default;
 $input-placeholder-color: mix($input-bg, $input-color) !default;
 
 $input-disabled-bg: mix($input-color, $input-bg, 10%) !default;
+$input-disabled-border-color: rgba($input-color, 0.15) !default;
+
+$form-select-disabled-color: mix($input-color, $input-bg) !default;
+$form-select-disabled-bg: $input-disabled-bg !default;
+$form-select-disabled-border-color: $input-disabled-border-color !default;
```
Same

Brieuc-brd commented:
Branch rebased and conflict resolved

Brieuc-brd commented:
@qsm-odoo 👋 
Little reminder :)

qsm-odoo reviewed (changes requested):
Disabled form-select before your branch:

![image](https://github.com/odoo/odoo/assets/10338094/d9090a64-2247-4e3f-9f68-52c6dcaa2b1c)

After:

![image](https://github.com/odoo/odoo/assets/10338094/b5aff19a-6b29-4cb9-a7d8-f755cda0b45f)


That's an example I found playing with the DOM by myself / wandering across apps to find problematic inputs in 17.0... which I did not find -> It would be nice if the commit message contained steps to reproduce the issues you are trying to fix in this stable version (17.0).

Also, range before:
![image](https://github.com/odoo/odoo/assets/10338094/dc150100-9d91-4696-aacf-9b0997fea1e6)

After: 
![image](https://github.com/odoo/odoo/assets/10338094/ca8d8a78-a2b9-409e-be1f-045cd9dedda4)

Is it wanted? Is this really needed as a fix? Should not that part be reserved for master? (and I actually preferred the "before" version).

Thanks :pray:

Brieuc-brd commented:
Hey @qsm-odoo 👋  Thanks for your review 👍 

> That's an example I found playing with the DOM by myself / wandering across apps to find problematic inputs in 17.0

I've fixed the `website_sale` dropdown and I also took the opportunity to make this dropdown work with the `multiple` attribute as the default dropdown (d16ed26de2d5aaaba7f300426070ea5b63bdab8d).

>  It would be nice if the commit message contained steps to reproduce the issues you are trying to fix in this stable version (17.0).

I've also reviewed the commits to split them to cover one issue per commit and added more details to reproduce these issues.

> Is it wanted? Is this really needed as a fix? Should not that part be reserved for master?

On second thought, this type of visual change should not be stable. I've removed it and kept only the color corrections.

I've also updated the PR description with examples for each problem solved 👍.

Can you take a look at it ? 

Thanks in advance !

qsm-odoo commented on addons/web/static/src/scss/bootstrap_review_frontend.scss:
```diff
@@ -236,3 +236,37 @@ $-color-for-gray-200-bg: adjust-color-to-background($body-color, $gray-200);
+    // (the color may be darker depending on the background color used).
+    background-clip: padding-box;
+
+    &:disabled:not([multiple]):not([size]:not([size="1"])) {
```
The `:not([size]:not([size="1"]))` part gives me a headache ^^ It this overriding a specific bootstrap rule? If not maybe `:where(:not([size]), [size="1"])` would be more readable and less forcing ?

Brieuc-brd replied:
`:where(:not([size]), [size="1"])` works well, thanks ! 👍

qsm-odoo commented on addons/web/static/src/scss/bootstrap_review_frontend.scss:
```diff
@@ -236,3 +236,37 @@ $-color-for-gray-200-bg: adjust-color-to-background($body-color, $gray-200);
+
+.form-control {
+    &:disabled, &[readonly] {
+        color: mix($input-color, $input-bg);
```
You use this at some places but this is equal to the value of `$form-select-disabled-color`, should not you uses it ?

Brieuc-brd replied:
Good catch ! I've updated the code to use this variable when necessary.

qsm-odoo commented on addons/website_sale/static/src/scss/product_configurator.scss:
```diff
@@ -85,6 +85,10 @@ select.form-select.css_attribute_select {
     background-repeat: no-repeat;
     max-width: 400px;
+
+    &:disabled:not([multiple]):not([size]:not([size="1"])) {
```
Is not there a way to share that selector somehow? (mixin?)

qsm-odoo commented on addons/web/static/src/scss/bootstrap_overridden_frontend.scss:
```diff
@@ -69,6 +69,11 @@ $input-disabled-border-color: rgba($input-color, 0.15) !default;
 
 $form-check-input-border: $border-width solid $gray-400 !default;
 $form-check-input-border-radius: 0 !default;
+$form-check-input-checked-color: color-contrast($primary, $body-color, $white)!default;
```
- Missing space to be consistent with the rest :wink: 
- Sure using `$primary` is the best here? Maybe `$component-active-bg` is better? (as it is also overridden in website?)

Brieuc-brd replied:
Indeed, I've updated the code 👍

qsm-odoo reviewed (changes requested):
Thanks for the split and examples, it always helps a lot :+1: I did not test yet, just a few comments of things to review/change IMO then I'll test and merge :pray:

Brieuc-brd commented on addons/web/static/src/scss/bootstrap_review.scss:
```diff
@@ -93,7 +93,7 @@
-.form-check, .form-select {
+.form-check, .form-select:where(:not(:disabled)) {
     @include o-field-pointer();
 }
```
Indeed, it makes sense, I've updated the code 👍 
https://github.com/odoo/odoo/pull/150886/files#diff-5c33398fe77e998d35e5772b25b414c7a0701292e61bd1500f817daccd3e11c3

Brieuc-brd commented:
Thanks for your review @qsm-odoo 🙏 
I've updated the branch, could you have a look at it ?

qsm-odoo reviewed (approved):
Thanks for the work again :+1: Hopefully no mistake anymore as it's more tricky than it seems :grimacing: 

I push-forced to:
- Make the last commit the first one: indeed without it, the previously first commit broke the /shop select when disabled as previously advertised, so best fix things in order:
![image](https://github.com/odoo/odoo/assets/10338094/fe31559f-c1ec-464b-82e6-dec6c13fa242)
- Review the mentioned apps in commits (they were very nice otherwise)
- As discussed on Discord: change the form-range component's track background to use the suggested value explained below (I also added a big comment in the code with our discussion):
![image](https://github.com/odoo/odoo/assets/10338094/3b80bfb5-3904-4664-98bb-1b1f15ba95b5)

(note: of course, don't make your inputs red please :hand_over_mouth:)

@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.
